src/object/cli_obj.c

```diff
@@ -180,9 +180,9 @@ obj_shard_open(struct dc_object *obj, unsigned int shard, unsigned int map_ver,
 		rc = dc_obj_shard_open(obj, oid, obj->cob_mode, obj_shard);
 		if (rc)
 			D_GOTO(unlock, rc);
-	}
 
-	if (rc == 0) {
+		*shard_ptr = obj_shard;
+	} else {
```

Comment:
this one looks incorrect to me, as now you only take one addref for both case that 1) first-time shard open, and 2) re-usable case when obj shard already opened.
This change will not cause failure, but will cause the obj_shard open cache actually cannot usable (cannot reuse the opened obj_shard), this may affect performance a little.

The original intent why take 2 times addref (one in dc_obj_shard_open  and another obj_shard_addref  below) is to make the opened obj_shard be reusable until obj_close().
below L187's addref will be released in obj_shard_open() corresponding obj_shard_close;
and first-time's addref in above dc_obj_shard_open, will be released in obj_close -> obj_layout_free(), then the obj_shard can be reused between obj open/close.

why you want to change it, did you observe mem leak for this case?

Reply:
hmm, yes, the layout is not released in some cases. But I probably can revert this change for the moment.  Actually we probably should not reset do_obj to NULL in obj_shard_decref(), so to reuse the cache, instead of messing with the refcount and open count.